### PR TITLE
APEXCORE-743 Killed container shown as running

### PR DESCRIPTION
**Pull request by sandeshh**
@PramodSSImmaneni @vrozov please review.

### CODE REVIEW

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (sanjaypujare) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: Instead of hardcoding 30 * 1000 can we make it configurable or at least define a const variable?

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (sanjaypujare) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: It might be better to populate pendingContainersToKill inside sendContainerAskToRM() where the last loop goes thru all of dnmgr.containerStopRequests and optionally issues a stopContainerAsync. At that point we can add the current containerIdStr  entry to pendingContainersToKill with the current-time. The logic is in one place so easier to understand (also possibly synchronized on containerStopRequests to prevent race conditions).

Comment (sanjaypujare) [reply engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: An alternative (more modular/OO) implementation could be to move this logic into the existing NMCallbackHandler class. Save the reference to NMCallbackHandler in StreamingAppMasterService (or get it from nmClient.getCallbackHandler) and move pendingContainersToKill in there. When NMCallbackHandler.onContainerStopped(ContainerId) is called remove the containerId from pendingContainersToKill and then only examine that list for containers that have been pending there for a long time.

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: Done.

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: Regarding 
1. sendContainerAskToRM is already overloaded, as kill request is sent to NM and not to RM, so adding more things to it will only complicate it further.
2. We can't move this logic to NMCallbackHandler. This PR is because onContainerStopped may not have been called.

Comment (sanjaypujare) [reply engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: 1. What do you mean overloaded? If the name is misleading we can rename it (it is private) to indicate what it is supposed to be doing including the new code in this PR suggested by me.

2. The end result is the same. When onContainerStopped is called the contaierIds are removed from pendingContainersToKill and if it is not called (the case you are saying) we will detect the containers that have been in pendingContainersToKill for too long (more than NODE_MANAGER_KILL_CONTAINER_TIMEOUT).

I suggest you reconsider

**Review by sanjaypujare — CHANGES_REQUESTED** [review]
suggest refactor to make it more modular

**Review by sanjaypujare — APPROVED** [review]
Looks ok

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: Is a separate logic block needed for this? Can't the addition to the pendingContainersToKill go into the previous if block.

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: Is there a specific need for refactoring this logic out of sendContainerAskToRM into a separate method. The sendContainerAskToRM method already seems to be handling all the different kinds of container requests.

Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: Method name was misleading, it was sending a request to both RM and NM. Now it is clearly separated and this also addresses @sanjaypujare's code review comments.

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: No need for a separate block, rolling this up.

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: Can you make it configurable by a system property. See bufferserver.server.Server.BACK_PRESSURE_ENABLED

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: Extra space before this line

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: Why can't the first for section remain in sendContainerAskToRM as that method is dealing with sending of all requests to RM.

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: done.

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: done.

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (pramodin) [reply engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: How about calling it processPendingContainers, see comment below on the method format.

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: greater or equal

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: How about a more appropriate name as this timeout is not a Node Manager setting but rather a fallback setting. You could just call it REMOVE_CONTAINER_TIMEOUT (after the RM call removeContainerRequest)

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: Done.

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: done.

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: Done.

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: REMOVE_CONTAINER_TIMEOUT = Integer.getInteger("org.apache.apex.nodemanager.kill.timeout", 30 * 1000);

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: Is not it subject to ConcurrentModificationException?

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: The order of the methods has changed

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: encapsulate logic allocatedContainer.stopRequested != -1 into AllocatedContainer:`allocatedContainer.isStopRequested()`

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: move to AllocatedContainer: `allocatedContainer.stop()'

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: The same as other comments. Also, move REMOVE_CONTAINER_TIMEOUT.

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: done.

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: done.

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: fixed.

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java
Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: Done.

**Comment by sandeshh** [comment]
test this please

**Comment by sandeshh** [comment]
test this please

**Comment by sandeshh** [comment]
@vrozov @PramodSSImmaneni All the review comments have been addressed, please review and merge.

**Comment by pramodin** [comment]
I am LGTM. I will wait for an LGTM from @vrozov and merge if needed.

**Review by vrozov — APPROVED** [review]

**Pull request merged**
